src/lib/storage/table.cpp

```diff
@@ -17,53 +19,103 @@
-  // Implementation goes here
-  return ChunkID{0};
-}
+ChunkID Table::chunk_count() const { return (ChunkID)_chunks.size(); }
 
 ColumnID Table::column_id_by_name(const std::string& column_name) const {
-  // Implementation goes here
-  return ColumnID{0};
+  auto search = _column_name_mapping.find(column_name);
+  DebugAssert(search != _column_name_mapping.end(), "Could not find column_name");
+  return search->second;
 }
 
-uint32_t Table::max_chunk_size() const {
-  // Implementation goes here
-  return 0;
-}
+uint32_t Table::max_chunk_size() const { return _max_chunk_size; }
 
-const std::vector<std::string>& Table::column_names() const {
-  throw std::runtime_error("Implement Table::column_names()");
-}
+bool Table::is_empty() { return _chunks.size() == 1 && _chunks.front()->size() == 0; }
 
-const std::string& Table::column_name(ColumnID column_id) const {
-  throw std::runtime_error("Implement Table::column_name");
-}
+const std::vector<std::string>& Table::column_names() const { return _column_names; }
 
-const std::string& Table::column_type(ColumnID column_id) const {
-  throw std::runtime_error("Implement Table::column_type");
-}
+const std::string& Table::column_name(ColumnID column_id) const { return _column_names.at(column_id); }
+
+const std::string& Table::column_type(ColumnID column_id) const { return _column_types.at(column_id); }
+
+Chunk& Table::get_chunk(ChunkID chunk_id) { return *_chunks.at(chunk_id); }
+
+const Chunk& Table::get_chunk(ChunkID chunk_id) const { return *_chunks.at(chunk_id); }
+
+void Table::compress_chunk(ChunkID chunk_id) {
+  DebugAssert(chunk_id < _chunks.size(), "ChunkID out of range");
+  const auto chunk = _chunks[chunk_id];
+  auto compressed_chunk = std::make_shared<Chunk>();
+  // create new empty chunk
+  _chunks.push_back(compressed_chunk);
 
-Chunk& Table::get_chunk(ChunkID chunk_id) { throw std::runtime_error("Implement Table::get_chunk"); }
+  // number of threads needed
+  auto n_columns = chunk->column_count();
+  std::vector<std::thread> threads(n_columns);
+  std::vector<std::shared_ptr<BaseSegment>> dictionary_segments(n_columns);
+  // to enable threads to access specific position in vector
+  dictionary_segments.resize(n_columns);
 
-const Chunk& Table::get_chunk(ChunkID chunk_id) const { throw std::runtime_error("Implement Table::get_chunk"); }
+  for (size_t segment = 0; segment < dictionary_segments.size(); ++segment) {
+    threads.push_back(std::thread([&] (size_t seg) {
+        std::string data_type = column_type((ColumnID) seg);
+        std::shared_ptr<BaseSegment> current_segment = chunk->get_segment((ColumnID) seg);
+        auto compressed_chunk = make_shared_by_data_type<BaseSegment, DictionarySegment>
+                        (data_type, current_segment);
+
+        dictionary_segments[seg] = compressed_chunk;
+    }, segment));
+  }
+
+  for (size_t thread=0; thread < threads.size(); thread++) {
+    if (threads[thread].joinable()) {
```

Comment:
this should be always true since all threads have been started and have not been joined yet. could be omitted

Reply:
I'd like that to be true as well. Sadly, not testing the condition here lead to segfaults in around 30% of our runs. From what I read the thread.join method fails, if the process is not yet started by the OS and just creating the thread in C++ does not guarantee that the OS already created the thread as well.

Reply:
ok. but what happens, if a thread is really not joinable at the moment of this if statement? since it is never tested again, the thread is never joined and it's not guaranteed that all segments are properly compressed when the method returns. this could lead to unexpected behaviour later on. if the struggle is real, it should mabe rather be 
`while(!threads[thread].joinable()){}
threads[thread].join();`

Reply:
That is a really good point actually. Thanks for catching the bug

Reply:
Well we found an actual bug via your comment in our program. 

The reason why we could not call `thread.join()` in our code was that we had a few default constructed threads in our `threads` vector by accident. The code that checks for `thread.joinable()` later skips those default constructed threads so nothing happens and the code works fine :D 

One of my funnier bugs recently. Thanks again!

Reply:
Fixed in next commit